lib/AST/ASTImporter.cpp

```diff
@@ -2417,21 +2423,33 @@ Decl *ASTNodeImporter::VisitFunctionDecl(FunctionDecl *D) {
   QualType T = Importer.Import(FromTy);
   if (T.isNull())
     return nullptr;
+  AlreadyImportedD = Importer.GetAlreadyImportedOrNull(D);
```

Comment:
Why do we need these lines? Didn't we check already the same thing in line 2398-2400?

Reply:
The same thing was checked but not on the same state. The state changes in between and the condition becomes true exactly here, not in the previous check (in the concrete case for the assert in `addHiddenDecl`). Maybe the last check can be enough, important is not to create a new `FunctionDecl*` when there is already one for this `D` argument. At the time when the ASTImporter::Import was called with this `D` the map did not yet contain an entry.

(More details: The assertion is triggered because such parameter `ParmVarDecl` objects are imported that are already part of another existing `FunctionDecl` that is imported from the same function as `D`. It looks like that this object is created by the import statement in line 2423.)

Reply:
I think that between these checks, different things happen. If only this QualType import is prone to do such a rouge import, I would only put a check here. Even if theoretically there is a chance that this happens at other places, we don't have an example yet.